File: knapsack.cpp
#include<iostream>

using namespace std;
int max_e=0;
void solve(int arr[],int sum,int i,int k,int n)
{
if(i>=n)
return  ;
if(max_e<sum && sum<=k)
max_e=sum;
if(sum+arr[i]>k)
solve(arr,sum,i+1,k,n);
else solve(arr,sum+arr[i],i,k,n);

}

int main()
{
int t,n,k,arr[2000];
cin>>t;
while(t--){
max_e=0;
cin>>n>>k;
for(int i=0;i<n;++i)
cin>>arr[i];
solve(arr,0,0,k,n);
cout<<max_e<<endl;
}
return 0;
}

File: knapsack.py
#!/usr/bin
mem=[0 for i in range(2001)]
def solve(arr,n,k):
	for i in range(1,k+1):
		for j in range(n):
			if int(arr[j])>i:continue
			mem[i]=max(mem[i],int(arr[j])+mem[i-int(arr[j])])			

t=int(input())
while t:
	n,k=input().split(" ")
	n=int(n)
	k=int(k)
	arr=input().split(" ")
	solve(arr,n,k)
	print(mem[k])
	mem=[0 for i in range(2001)]
	t-=1
